Sets up simulated GCS bucket directory browsing; fills in blob metadata

etanshaul opened the pull request:
This PR implements directory browsing in the GCS explorer. It also fills in the remaining blob metadata.

In the backend, there is no concept of "directories" within a GCS bucket. Everything is flattened and a blob just has a name and metadata. Therefore, simulation of directories was necessary to create the desired UX (a file browser-like experience). For well written writeup on this, see https://github.com/csells/BucketHelperSample

Note: this PR does not add any visual polish to the browser. Subsequent PRs will address look and feel / icons. Also remaining is implementation of the toolbar (with upload/download/refresh) and a breadcrumb navigator.

File browser:
![image](https://user-images.githubusercontent.com/1735744/30072306-dfdc4bb8-9237-11e7-848e-2c3fa7030fc8.png)

Double clicking a directory, reloads the display and shows the contents of the clicked directory:
![image](https://user-images.githubusercontent.com/1735744/30072329-f61df1d8-9237-11e7-9fff-a7c2373d0191.png)

etanshaul commented:
@nkibler, @patflynn you can start taking a look. Still need to update and add more tests.

etanshaul commented:
Note, for this to work, https://github.com/GoogleCloudPlatform/google-cloud-intellij/pull/1624 (or some similar solution) will need to be merged in first.

etanshaul commented:
@nkibler @patflynn this is ready for review. PTAL.

nkibler commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanel.java:
```diff
@@ -17,36 +17,69 @@
+// TODO(eshaul) implement toolbar panel and breadcrumb browser
 final class GcsBucketContentEditorPanel {
 
+  private Bucket bucket;
```
final?

etanshaul replied:
done

nkibler commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanel.java:
```diff
@@ -17,36 +17,69 @@
+  }
+
+  // TODO(eshaul) should be done asynchronously and show loader in the UI
+  private List<Blob> getBlobsAtLevel(String prefix) {
```
I'm not sure what "level" refers to here. Maybe this should be renamed to `getBlobsForPrefix()`?

etanshaul replied:
I updated the name to `getBlobsStartingWith` and added javadoc to explain whats going on. Let me know if its more clear now.

nkibler commented on google-cloud-tools-plugin/testSrc/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanelTest.java:
```diff
@@ -36,51 +40,97 @@
+  @Mock private Blob directoryBlob;
+  @Mock private Blob binaryBlob;
+
+  private BiMap<Integer, String> indexToColName =
```
Can these be `static final`?

etanshaul replied:
done

nkibler commented on google-cloud-tools-plugin/testSrc/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanelTest.java:
```diff
@@ -36,51 +40,97 @@
     bucketVirtualFile = GcsTestUtils.createVirtualFileWithBucketMocks();
+
+    when(directoryBlob.isDirectory()).thenReturn(true);
+    when(directoryBlob.getName()).thenReturn("my_directory");
```
Maybe pull these into static constants so writing tests below don't require you to read the `setUp()` method every time to remember what these values are?

etanshaul replied:
done

etanshaul commented on common-test-lib/src/main/java/com/google/cloud/tools/intellij/testing/TimeZoneRule.java:
```diff
@@ -0,0 +1,47 @@
+import org.junit.runners.model.Statement;
+
+/** A custom {@link TestRule} for setting a time zone in unit tests. */
+public class TimeZoneRule implements TestRule {
```
done

etanshaul commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/gcs/GcsBlobTableModel.java:
```diff
@@ -0,0 +1,115 @@
+  }
+
+  Blob getBlobAt(int index) {
+    return blobs.get(index);
```
hmm. if blobs is empty this method will still throw an IndexOutOfBounds excpetion. What about throwing a runtime exception if blobs is null?

etanshaul commented:
@nkibler ptal. Addressed your comments. Also added some more tests.

Regarding your comment about `getBlobAt` throwing an NPE, I updated to throw a RuntimeException if blobs are null. let me know what you think.

nkibler commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanel.java:
```diff
@@ -60,22 +60,32 @@ public void mouseClicked(MouseEvent event) {
   }
 
+  /**
+   * In the GCS backend there are no directories, just blobs with names. Implicit in the name,
```
nit: This seems like a supporting paragraph that should exist under a summary fragment for this method. Consider adding a summary fragment like:

```java
/**
 * Returns the list of {@link Blob Blobs} that start with the given prefix.
 *
 * <p>In the GCS backend...
 */
```

etanshaul replied:
done

nkibler commented on google-cloud-tools-plugin/testSrc/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanelTest.java:
```diff
@@ -47,23 +47,32 @@
+  @Mock private Blob nestedBlob;
 
-  private BiMap<Integer, String> indexToColName =
+  private static final String DIR_NAME = "my_directory";
```
nit: Can you move these static variables above the instance variables to maintain a logical ordering?

etanshaul replied:
done

nkibler commented on google-cloud-tools-plugin/testSrc/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanelTest.java:
```diff
@@ -47,23 +47,32 @@
+  private static final String NESTED_BLOB_FULL_NAME = "dir1/dir2/nested_blob.zip";
+  private static final String NESTED_BLOB_NAME = "nested_blob.zip";
+
+  private static final BiMap<Integer, String> indexToColName =
```
nit: Use CONSTANT_CASE for these, now that they're constants.

etanshaul replied:
done
